Refactor tlc saving in channel state and fix a few bugs

contrun opened the pull request:
This PR refactored TLC fields in channel state. We will save all the tlcs in a unified tlcs field. This make it more convenient to see when a tlc is added/removed. When refactoring this I also found a few bugs. Below are the changes in this PR.

1. A notification for local commitment signed `LocalCommitmentSigned` is added. This may be used to notify watch tower that some transactions are now broadcastable by the peer.
2. We now have complete information for when a tlc is added/removed.
3. Previously commitment number is updated on `RevokeAndAck` from the peer received. This effectively makes local/remote commitment number always the same. We should instead update commitment numbers on `CommitmentSigned` messages sent or received. This PR fixes that.
4. The commitment number used to obtain remote commitment point should always be local commitment number (instead of remote commitment number). This asymmetry takes some time to understand, and previously we implemented in the wrong way. A remote commitment point is only obtained when `RevokeAndAck` is received from the remote party. And `RevokeAndAck` is sent only when the remote party receives a `CommitmentSigned` message. This is exactly when the local commitment transaction is updated.

contrun committed:
Save all tlcs to channel state

Add notification on commitment transaction signed locally

Fix initial commitment number

Bundle local and remote commitment numbers in channel state

Bundle offering and received tlc ids in channel state

Use enum for TLC id

Use a unified field to represent all tlcs in channel state

Use btree map instead of hash map for tlcs

Refactor add tlc

Greatly simplify tlc handling

Remove two subtle bugs due to tlc data structure change

Fix json serde error because keys in objects must be strings

Refactor get active tlc implementation

Remove outdated code

Fix balance calculation in removing tlc

contrun committed:
Update remote commitment number on CommitmentSigned (not RevokeAndAck)

Commitment numbers of each parties are independent. Updating remote
commitment number after received RevokeAndAck effectively means we are
using the same commitment number.

contrun committed:
Use local commitment number for remote commitment point

A remote commitment point is only obtained when RevokeAndAck is
received from the remote party. And RevokeAndAck is sent only when the
remote party receives a CommitmentSigned message. This is exactly when
the local commitment transaction is updated.

chenyukang commented on src/ckb/channel.rs:
```diff
@@ -1590,34 +1645,183 @@ impl ChannelActorState {
+                        return Ok(current.clone());
+                    }
+                    Some((removed_at, remove_reason)) => {
+                        error!("Trying to remove an already removed tlc {:?} (removed at {:?}) with different reason: {:?} (old reason {:?})", current, removed_at, reason, remove_reason);
```
this branch seems we have `removed_at != commitment_numbers` or  `reason != remove_reason`,
but the error message implies only the different `reason`.

chenyukang commented on src/ckb/channel.rs:
```diff
@@ -1590,34 +1645,183 @@ impl ChannelActorState {
+                } else {
+                    self.to_remote_amount += current.tlc.amount;
+                }
+                current.removal_info = Some((commitment_numbers, reason));
```
is it better we remove the current entry from `self.tlcs`?
otherwise `tlcs` will keep growing, and we need to serialize and store this field.

contrun replied:
We need this historical tlc information to construct valid revocation transaction. Imagine that the adversary broadcasts an older commitment transaction, and we want to revoke this. In order to do so, we will need to fill in all the tlcs that are committed commitment transaction in the witness. This is the reason why we need to keep tlcs forever. Of course, we can also save the witnesses for commitment transaction (just hook into this event https://github.com/nervosnetwork/cfn-node/blob/c1c4557a15d6457b23008328e1fc5cdf82e9611b/src/ckb/network.rs#L152C5-L152C27). I personally prefer the latter method. But I haven't think carefully about the tradeoffs. I think it is not such so important to optimize this yet.

chenyukang commented on src/ckb/channel.rs:
```diff
@@ -1590,34 +1645,183 @@ impl ChannelActorState {
+        }
+    }
+
+    pub fn get_channel_parameters(&self, local: bool) -> &ChannelParametersOneParty {
```
seems unused function here.

chenyukang commented on src/ckb/channel.rs:
```diff
@@ -1590,34 +1645,183 @@ impl ChannelActorState {
+        };
+        self.tlcs.insert(tlc.id, detailed_tlc);
+        if tlc.is_offered() {
+            self.to_local_amount -= tlc.amount;
```
can we add an `blance_check` function, which will make sure `to_local_amount + to_remote_amount + pending tlcs amount == initial channel amount`, so that we can catch error whenever unexpected thing happened.

contrun replied:
Added in https://github.com/nervosnetwork/cfn-node/pull/72/commits/19d43491f9c5ccf660131c106060b9117adb09d3

chenyukang commented on src/ckb/channel.rs:
```diff
@@ -1751,25 +1925,163 @@ impl ChannelActorState {
+                // We may have inconsistent view of the active tlcs.
+                // This should not happen.
+                panic!(
+                    "Trying to obtain list of active received tlcs while remote has not confirmed our latest commitment transaction, should have verified we are not in this state before calling this function.",
```
it's better to use multiple lines.

chenyukang commented on src/ckb/channel.rs:
```diff
@@ -2751,8 +3055,12 @@ impl ChannelActorState {
+    pub fn build_commitment_tx(&self, local: bool) -> (TransactionView, [u8; 32], Vec<u8>) {
+        let version = self.get_current_commitment_number(local);
+        debug!("Building {} commitment transaction #{} with local commtiment number {} and remote commitment number {}", if local { "local" } else { "remote" }, version,
+    self.get_local_commitment_number(), self.get_remote_commitment_number());
```
bad indentation.

contrun committed:
Add channel balance invariant check

quake commented on src/ckb/channel.rs:
```diff
@@ -1751,25 +1945,163 @@ impl ChannelActorState {
+                Some(_) => None,
+                // This tlc is still active.
+                None => Some(detailed_tlc),
+            })
```
suggest to add an `is_removed` fn to DetailedTlcInfo and change the code to:
```suggestion
            .filter(|detailed_tlc| !detailed_tlc.is_removed() && detailed_tlc.tlc.is_received() )
```

quake commented on src/ckb/channel.rs:
```diff
@@ -1751,25 +1945,163 @@ impl ChannelActorState {
+                Some(_) => None,
+                // This tlc is still active.
+                None => Some(detailed_tlc),
+            })
```
suggest to add an `is_removed` fn to DetailedTlcInfo and change the code to:
```suggestion
            .filter(|detailed_tlc| !detailed_tlc.is_removed() && detailed_tlc.tlc.is_offered() )
```

quake commented on src/ckb/channel.rs:
```diff
@@ -1751,25 +1945,163 @@ impl ChannelActorState {
+    pub fn get_active_tlcs(&self) -> impl Iterator<Item = &DetailedTLCInfo> {
+        self.get_active_offered_tlcs()
+            .into_iter()
+            .chain(self.get_active_received_tlcs())
```
```suggestion
        self.tlcs
            .values()
            .filter(|detailed_tlc| !detailed_tlc.is_removed())
```

quake commented on src/ckb/channel.rs:
```diff
@@ -1751,25 +1945,163 @@ impl ChannelActorState {
+        self.get_active_tlcs()
+            .into_iter()
+            .map(|tlc| tlc.tlc.amount)
+            .sum::<u128>()
```
```suggestion
        self.get_active_tlcs()
            .map(|tlc| tlc.tlc.amount)
            .sum()
```

quake commented on src/ckb/channel.rs:
```diff
@@ -1202,6 +1128,144 @@ impl From<Transaction> for FundingTxInput {
+}
+
+impl CommitmentNumbers {
+    pub fn new(local: u64, remote: u64) -> Self {
```
I think this fn is useless because there are no scenarios need to set these 2 values directly.

quake commented on src/ckb/channel.rs:
```diff
@@ -1202,6 +1128,144 @@ impl From<Transaction> for FundingTxInput {
+    Received(u64),
+}
+
+// Required for using TLCId as key in BTreeMap.
```
I don't think BTreeMap has the `Hash` trait bound requirement for the key, could you double check it?

contrun replied:
Yes. You're right. I removed it.

quake commented on src/ckb/channel.rs:
```diff
@@ -1751,25 +1925,163 @@ impl ChannelActorState {
+    pub fn get_active_received_tlc_with_pubkeys(
+        &self,
+    ) -> impl Iterator<Item = (&DetailedTLCInfo, Pubkey, Pubkey)> {
+        self.get_active_received_tlcs().into_iter().map(move |tlc| {
```
just use map:
```suggestion
        self.get_active_received_tlcs().map(|tlc| {
```

quake commented on src/ckb/channel.rs:
```diff
@@ -1751,25 +1925,163 @@ impl ChannelActorState {
+    pub fn get_active_offered_tlc_with_pubkeys(
+        &self,
+    ) -> impl Iterator<Item = (&DetailedTLCInfo, Pubkey, Pubkey)> {
+        self.get_active_offered_tlcs().into_iter().map(move |tlc| {
```
just use map:
```suggestion
        self.get_active_offered_tlcs().map(|tlc| {
```

quake commented on src/ckb/channel.rs:
```diff
@@ -1751,25 +1925,163 @@ impl ChannelActorState {
+                self.get_active_offered_tlc_with_pubkeys()
+                    .into_iter()
+                    .map(|(tlc, local, remote)| (tlc.clone(), local, remote))
+                    .collect::<Vec<_>>(),
```
```suggestion
                self.get_active_received_tlc_with_pubkeys()
                    .map(|(tlc, local, remote)| (tlc.clone(), local, remote))
                    .collect::<Vec<_>>(),
                self.get_active_offered_tlc_with_pubkeys()
                    .map(|(tlc, local, remote)| (tlc.clone(), local, remote))
                    .collect::<Vec<_>>(),
```

quake commented on src/ckb/channel.rs:
```diff
@@ -1751,25 +1925,163 @@ impl ChannelActorState {
+                for tlc in received_tlcs.iter_mut().chain(offered_tlcs.iter_mut()) {
+                    // Need to flip these fields for the counterparty.
+                    tlc.0.tlc.flip_mut();
+                }
```
```suggestion
                for (tlc, _, _) in received_tlcs.iter_mut().chain(offered_tlcs.iter_mut()) {
                    // Need to flip these fields for the counterparty.
                    tlc.tlc.flip_mut();
                }
```

quake commented on src/ckb/channel.rs:
```diff
@@ -1751,25 +1925,163 @@ impl ChannelActorState {
+                }
+                (offered_tlcs, received_tlcs)
+            };
+            a.sort_by(|x, y| u64::from(x.0.tlc.id).cmp(&u64::from(y.0.tlc.id)));
+            b.sort_by(|x, y| u64::from(x.0.tlc.id).cmp(&u64::from(y.0.tlc.id)));
```
BTreeMap is sorted, I don't think we need sort again here. btw, we need unit test for this fn.

quake commented on src/ckb/channel.rs:
```diff
@@ -1590,34 +1645,183 @@ impl ChannelActorState {
+        };
+        self.tlcs.insert(tlc.id, detailed_tlc);
+        if tlc.is_offered() {
+            self.to_local_amount -= tlc.amount;
```
@chenyukang a reminder,  we need to check `to_local_amount / to_remote_amount - tlc.amount > min_amount` here before inserting.

contrun committed:
Refactor tlc structure to reflect on the asymmetry nature of commitment tx

contrun committed:
Update state only on RevokeAndAck message sent or received

contrun committed:
Refactor detailed tlc struct fields

contrun committed:
Refactor get active offered, received tlcs

contrun committed:
Save both commitment numbers for tlc

contrun committed:
Fix updating tlc info on revoke_and_ack

contrun committed:
Fix wrong state update in add tlc

contrun committed:
Fix calculating time locked value in commitment tx

contrun committed:
Remove a few useless functions/struct

contrun committed:
Document a potential bug for tlc created_at field

contrun committed:
Add test to ensure balance is corrent

wip

contrun committed:
Move all state changes away from commitment signed

contrun committed:
Add more channel balance checks

contrun committed:
Show channel balance in list_channels rpc

contrun committed:
Remove ChannelReadyFlags::AWAIT_REMOTE_REVOKE

contrun committed:
Use match to avoid evaluation in shutdown

https://stackoverflow.com/questions/56726571/why-choosing-unwrap-or-else-over-unwrap-or
Using unwrap_or will always evaluate the argument.
This is undesirable because we may send the message twice here.

contrun committed:
Add more debug info for shutdown balance
